File: 1029/C++/main.cpp
// https://www.urionlinejudge.com.br/judge/en/problems/view/1019
#include <iostream>

typedef struct {
  int calls;
  int ones;
} FibCount;

FibCount fib_count(int x) {
  FibCount a = { .calls = 1, .ones = 0 };
  FibCount b = { .calls = 1, .ones = 1 };
  FibCount c;

  if (x <= 0) return a;
  if (x == 1) return b;

  for (int i = 1; i < x; i++) {
    c.calls = a.calls + b.calls + 1;
    c.ones = a.ones + b.ones;
    a = b;
    b = c;
  }

  b.calls--;
  return b;
}

int main() {
  int n, x;
  FibCount result;
  std::cin >> n;

  while (n-- > 0) {
    std::cin >> x;
    result = fib_count(x);
    std::cout << "fib(" << x << ") = " << result.calls << " calls = " << result.ones << std::endl;
  }

  return 0;
}

File: 1029/Haskell/main.hs
-- https://www.urionlinejudge.com.br/judge/en/problems/view/1029

data Fib = Fib 
  { fibValue :: Integer
  , fibCalls :: Integer
  , fibOnes :: Integer
  }

fibNext (Fib a1 c1 o1) (Fib a2 c2 o2) = Fib (a1 + a2) (c1 + c2 + 1) (o1 + o2)

fibCount = Fib 0 1 0 : Fib 1 1 1 : zipWith fibNext fibCount (tail fibCount)

showFib n = do
  let fn = fibCount !! read n
  "fib(" ++ n ++ ") = " ++ show (fibCalls fn - 1) ++ " calls = " ++ show (fibOnes fn)

main = interact (unlines . map showFib . tail . lines)
